File: Day Challenge/20230224 Capacity To Ship Packages Within D Days/Capacity To Ship Packages Within D Days.py
class Solution(object):
    def shipWithinDays(self, weights, days):
        left = max(weights)
        right = sum(weights)

        while left < right:
            mid = (left + right) // 2
            day = 1
            total_weight = 0

            for weight in weights:
                if total_weight + weight > mid:
                    day += 1
                    total_weight = 0
                total_weight += weight

            if day > days:
                left = mid + 1
            else:
                right = mid

        return left

File: Day Challenge/2023-1/20230224 Capacity To Ship Packages Within D Days/Capacity To Ship Packages Within D Days.js
/**
 * @param {number[]} weights
 * @param {number} days
 * @return {number}
 */

const shipWithinDays = function (weights, days) {
  let left = Math.max(...weights);
  let right = weights.reduce((acc, cur) => acc + cur, 0);

  while (left < right) {
    let mid = Math.floor((left + right) / 2);
    let day = 1;
    let sum = 0;
    weights.forEach((weight) => {
      if (sum + weight > mid) {
        day++;
        sum = 0;
      }
      sum += weight;
    });
    if (day > days) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }

  return left;
};
